packages/grid/x-data-grid-premium/src/concept/DataGrid.tsx

```diff
@@ -0,0 +1,606 @@
+      minY: y,
+      maxY: Math.min(y + BLOCK_ROW_SIZE - 1, this.rows.length),
+      didMount: false,
+      didRender: false,
+      element,
+      root,
+    }
+
+    this.blockTree.insert(block)
+
+    debug('blocks:', this.blockTree.all().length)
+  }
+
+  updateCanvas = () => {
+    const ctx = this.canvas.current!.getContext('2d')!
+
+    const w = this.canvas.current!.width
+    const h = this.canvas.current!.height
+
+    ctx.clearRect(0, 0, w, h)
+    ctx.fillStyle = 'rgba(255, 255, 255, 0.1)'
+    ctx.fillRect(0, 0, w, h)
+
+    const f = 0.1
+    const cellHeight = this.rowHeight * f * 0.15
+    const cellWidth = Math.round(this.columnWidth * f * 0.4)
+
+    ctx.lineWidth = 0.125
+
+    const doRect = (range: Range, op = ctx.strokeRect.bind(ctx)) => {
+      const x = range.minX * cellWidth
+      const y = range.minY * cellHeight
+      const w = (range.maxX + 1) * cellWidth - x - ctx.lineWidth
+      const h = (range.maxY + 1) * cellHeight - y - ctx.lineWidth
+      op(x + 0.5, y + 0.5, w, h)
+    }
+
+    const blocks = this.blockTree.all()
+    for (const block of blocks) {
+      ctx.strokeStyle = block.didRender ? 'green' : 'blue'
+      doRect(block)
+    }
+
+    ctx.lineWidth = 0.25
+    ctx.strokeStyle = 'red'
+    doRect(this.displayRange)
+
+    ctx.strokeStyle = 'yellow'
+    ctx.fillStyle = 'rgba(255, 255, 0, 0.1)'
+    doRect(this.screenRange)
+    doRect(this.screenRange, ctx.fillRect.bind(ctx))
+
+
+    this.canvasFrameId = requestAnimationFrame(this.updateCanvas)
+  }
+
+  render() {
+    return (
+      <div>
+        <div
+          className='grid'
+          style={{
+            '--row-height': `${this.rowHeight}px`,
+            '--header-height': `${this.rowHeight}px`,
+          } as React.CSSProperties}
+        >
+          <div ref={this.scrollVertical} className='fake-scroll-vertical'>
+            <div className='fake-scroll-vertical-content'></div>
+          </div>
+
+          <div ref={this.container} className='container'>
+            <div className='header'>
+              {this.columns.map((c) =>
+                <div className='column' key={c.field}>{c.headerName ?? c.field}</div>
+              )}
+            </div>
+
+            <div ref={this.content} className='content'></div>
+          </div>
+        </div>
+        <br/>
+        <canvas ref={this.canvas} width='700' height='400' />
+      </div>
+    );
+  }
+}
+
+function Block({ instance, rowIndex, columnIndex, onDidMount, onDidUnmount }: {
+  instance: DataGrid,
+  rowIndex: number;
+  columnIndex: number;
+  onDidMount?: Function;
+  onDidUnmount?: Function;
+}) {
+  React.useEffect(() => {
+    onDidMount?.()
+    return onDidUnmount as any
+  }, [])
+
+  return (
+    <React.Fragment>
+      {instance.rows.slice(rowIndex, rowIndex + BLOCK_ROW_SIZE).map((row, i) =>
+        <div className='row' key={i}>
+          {instance.columns.slice(columnIndex, columnIndex + BLOCK_COL_SIZE).map((column, j) =>
```

Comment:
This is interesting, the column header is not virtualized here.
I was wondering how would this approach scale with a large number of columns.

I tried to apply this idea to our current virtualization engine in https://github.com/mui/mui-x/pull/9589
Here is a demo with 1000 columns: https://deploy-preview-9589--material-ui-x.netlify.app/x/react-data-grid/virtualization/#column-virtualization
1. The initial render is laggy since we need to render 1000 columns at once.
2. The horizontal scrolling feels pretty good - the column header stays in sync with the content
3. Interactions with column header are super slow - e.g. keyboard navigation, column menu, sorting.
This is probably due to the fact that we rerender the whole column header on state change. We could optimize it with React.memo and only rerender a single cell at once, but this could be a bottleneck.

Overall, a non-virtualized column header feels risky to me.
What do you think?
Did you explore the idea of column header virtualization?

Reply:
I didn't implement it to keep things simple, but there isn't any issue with doing header virtualization and sticky headers.

However, we might want to disable header virtualization under a certain threshold (something like 100?). Rendering 100 column headers upfront isn't an issue, and we avoid some costs when scrolling.